### PR TITLE
Use the Dolmen identifiers for constructors and ADT names

### PR DESCRIPTION
**Pull request by Halbaroth**
We don't manage correctly incremental type declarations. For instance the following input is accepted by the SMT-LIB standard:
```smt2 
(set-logic ALL)
(push 1)
(declare-datatype t
  ((c1 (d1 Int)) (c2)))
(pop 1)
(declare-datatype t
  ((c2 (d1 Int)) (c1)))
(declare-const x t)
(declare-const y t)
(assert ((_ is c1) x))
(assert ((_ is c1) y))
(assert (distinct x y))
(check-sat)
```
(Notice that we switch the constructors `c1` and `c2` in our second declaration of the type `t`). AE answers `unknown` on `next` but it answers `unsat` if we remove the first declaration of `t`.

This PR fixes the issue for the new frontend only. So we got `unsat` with `--frontend dolmen` but we still got `unknown` with `--frontend legacy`.

The solution consists in using the unique identifiers produced by `Dolmen` for the constructors, destructors and names of ADTs. To reduce as much as possible the amount of modification done in the legacy typechecker, I wrapped Dolmen's identifiers in a new module `Uid`. This module agrees with `Hstring` if we use the constructors `of_string` or `of_hstring` and it uses Dolmen's identifiers if we use the constructor `of_dolmen`. Of course, we'll get rid of this module as soon as we remove the legacy frontend. 

I didn't use a unique identifier for the constructor of builtin enumeration types. This case will be fixed after removing the legacy frontend.

This PR is supposed to be merged before #1093

### CODE REVIEW

File: src/lib/structures/uid.ml
Comment (bclement-ocp) [inline src/lib/structures/uid.ml]: Why use a separate `Unique` constructor with our own tag rather than directly use the Dolmen identifier?

File: src/lib/structures/ty.ml
Comment (bclement-ocp) [inline src/lib/structures/ty.ml]: I believe this is required for the legacy typechecker not to confuse constructors for distinct record types ; it should be moved to the legacy typechecker.ml rather than removed entirely.

File: src/lib/structures/ty.ml
Comment (bclement-ocp) [inline src/lib/structures/ty.ml]: We should instead use `DStd.Ty.Const.unit`.

File: src/lib/structures/fpa_rounding.ml
Comment (bclement-ocp) [inline src/lib/structures/fpa_rounding.ml]: Why use hstring rather than dolmen IDs here?

File: src/lib/structures/uid.ml
Comment (bclement-ocp) [inline src/lib/structures/uid.ml]: Nit: calling them `Fake` is confusing; what's fake about them?

**Review by bclement-ocp — COMMENTED** [review]
I did not look thoroughly through the minutiæ of all mechanical changes from `Hstring` → `Uid` but the change looks correct overall. I do have some minor questions, and something needs to be done for record constructors which I think are broken in the legacy typechecker by this PR.

File: src/lib/structures/ty.ml
Comment (Halbaroth) [reply src/lib/structures/ty.ml]: I don't know `Dolmen` already contains a unit constant. I'll change it.

File: src/lib/structures/uid.ml
Comment (Halbaroth) [reply src/lib/structures/uid.ml]: It's fake unique identifier. It's not unique but I can rename it into `Hstring`.

File: src/lib/frontend/d_cnf.ml
Comment (bclement-ocp) [inline src/lib/frontend/d_cnf.ml]:
```suggestion
type id = Id : 'a DE.id -> id[@@unboxed]
```

File: src/lib/structures/uid.ml
Comment (Halbaroth) [reply src/lib/structures/uid.ml]: Of course I tried to use the `Dolmen identifier` directly but I got a regression in our non-regression test suite. I should investigate.

Comment (Halbaroth) [reply src/lib/structures/uid.ml]: Ok, I tried again with the original index produced by Dolmen. Now I got five regressions and it seems that Alt-Ergo ignores the timeout for them (`make runtest` never returns). Actually, if I run `dune exec -- alt-ergo tests/adts/match_in_terms.ae`, I cannot kill the processus with `CTRL+C`. I have to type `pkill alt-ergo`.

Comment (Halbaroth) [reply src/lib/structures/uid.ml]: I suspect an issue with `DE.Id.compare` or `DE.Id.equal`.

Comment (Halbaroth) [reply src/lib/structures/uid.ml]: Uhm they seems to be reasonable: 
```ocaml 
  let compare v v' = compare v.index v'.index

  let equal v v' = compare v v' = 0
```
very strange ...

Comment (Halbaroth) [reply src/lib/structures/uid.ml]: Thanks to Vincent, I got the bug with `gdb`. The issue comes from the fact that `Dolmen` identifiers can be infinite values and the polymorphic equality loops infinitely on them. I wrote another PR to remove as much as possible polymorphic comparison in lists.

File: src/lib/structures/fpa_rounding.ml
Comment (Halbaroth) [reply src/lib/structures/fpa_rounding.ml]: When I wrote this PR  I didn't do it because I didn't want to have to manage unique identifiers in the legacy typechecker. Basically, if we use real unique identifiers and hstring identifiers, we have to ensure that the legacy typechecker won't recreate ADTs' constructors for builtin enums (and it does...). 

Finally I found another way to use `dolmen` identifiers here. I modified the comparison functions of `Uid`. Now if we compare a `Hstring` and a `Dolmen` identifiers, we transform the `Dolmen` identifier into string.

Comment (bclement-ocp) [reply src/lib/structures/fpa_rounding.ml]: > the legacy typechecker won't recreate ADTs' constructors for builtin enums (and it does...).

Ah, I guess this is the issue of having different names for the AE and SMT enums, got it.

> Now if we compare a Hstring and a Dolmen identifiers, we transform the Dolmen identifier into string.

Is this not re-introducing the dependency on string values that this PR is trying to eliminate?

Comment (Halbaroth) [reply src/lib/structures/fpa_rounding.ml]: Now the `Dolmen` frontend doesn't produce `Uid.t` of type `Hstring` anymore (but for abstract types, I'll change that). It means we always compare `Dolmen` identifiers here (for constructors, destructors and ADTs' names).

**Comment by Halbaroth** [comment]
The PR is ready for a second pass. I think we should test the legacy frontend with `fpa` on Marvin before merging it.

File: src/lib/frontend/d_cnf.ml
Comment (bclement-ocp) [inline src/lib/frontend/d_cnf.ml]: I think you want `?id` here and `~id` when `fresh_ty` is called.

(Currently `fresh_ty` has a `?id:Uid.t option` optional argument, simply using `?id` would make it a `?id:Uid.t` argument)

File: src/lib/frontend/d_cnf.ml
Comment (bclement-ocp) [inline src/lib/frontend/d_cnf.ml]: This looks a bit weird; aren't we building a second Dolmen identifier with the same name as an existing one here?

File: src/lib/structures/uid.ml
Comment (bclement-ocp) [inline src/lib/structures/uid.ml]: Why not `DE.Id.print`?

**Review by bclement-ocp — COMMENTED** [review]
Yeah on paper the changes look correct but we definitely should check that there are no regressions. I am a bit lost by the many small changes now; I will take a deeper look after the weekend. I think there are many unneeded string/Hstring conversions now but we can clean that up later.

File: src/lib/structures/uid.ml
Comment (Gbury) [reply src/lib/structures/uid.ml]: Yeah, dolmen identifiers should never be compared using the polymorphic identity: besides being slow (compared to the comparison defined by dolmen), it may loop forever (because the `Type` identifiers loops on itself), or may fail at runtime (because ids contain tags that can contain arbitrary data such as functions).

File: src/lib/frontend/d_cnf.ml
Comment (Halbaroth) [reply src/lib/frontend/d_cnf.ml]: Indeed, I thought it was correct but we duplicate the identifiers here. I fixed the issue by creating the constructors and the Dolmen type in `Fpa_rounding`.

**Comment by Halbaroth** [comment]
I rebase this PR on #1117 to remove some polymorphic comparisons again ;)

File: src/lib/structures/uid.ml
Comment (bclement-ocp) [inline src/lib/structures/uid.ml]: Isn't this a complicated way of writing `String.equal (show u1) (show u2)`? The complexity should be the same since `Hstring.make` requires a linear scan.

File: src/lib/structures/uid.ml
Comment (bclement-ocp) [inline src/lib/structures/uid.ml]: Do we need the `Hstring` comparison here? Otherwise `String.compare (show u1) (show u2)` might be sufficient (complexity should be the same since `Hstring.make` requires a linear scan).

File: src/lib/structures/fpa_rounding.mli
Comment (bclement-ocp) [inline src/lib/structures/fpa_rounding.mli:54]: :+1: 

Do we still need the various string conversion functions with this?

File: src/lib/structures/fpa_rounding.ml
Comment (bclement-ocp) [inline src/lib/structures/fpa_rounding.ml]: Note: I think this could be replaced with `DStd.Expr.Ty.roundingMode`. I don't think it is entirely trivial to do so, but might be worth a note in a comment.

**Review by bclement-ocp — COMMENTED** [review]
Still a couple questions but overall I think this is ready to merge once rebased & if the benches are green.

File: src/lib/structures/uid.ml
Comment (Halbaroth) [reply src/lib/structures/uid.ml]: I thought it was a little bit faster because we compare integers if these `Hstring` already exist but we have to compute the hash of the strings, which may be linear in the length of them.

File: src/lib/structures/fpa_rounding.mli
Comment (Halbaroth) [reply src/lib/structures/fpa_rounding.mli:54]: I don't think so

File: src/lib/structures/fpa_rounding.ml
Comment (Halbaroth) [reply src/lib/structures/fpa_rounding.ml]: I didn't the enum type is shipped with `Dolmen`. We can replace it easily. It's exactly the same code.

Comment (Halbaroth) [reply src/lib/structures/fpa_rounding.ml]: Uhm actually `roundingMode` is the SMT-LIB roundingMode, so it's not so obvious. I add a note.

File: src/lib/structures/fpa_rounding.mli
Comment (Halbaroth) [reply src/lib/structures/fpa_rounding.mli:54]: We still need them in both legacy and new frontend. In the new frontend, we use these functions in the typechecker to inject the AE identifiers into SMT-LIB identifiers. 

We also use them in `Expr` because we need to convert a `Hstring` identifier into `Fpa_rounding.rounding_mode`. 

I'm sure we can remove some of them after getting rid of the legacy frontend.

File: src/lib/structures/fpa_rounding.ml
Comment (bclement-ocp) [reply src/lib/structures/fpa_rounding.ml]: The `RoundingMode` type from AE (SMT-LIB flavor) should be identical to the SMT-LIB RoundingMode, we made sure with @Stevendeo that the constructors are the same.

Comment (Halbaroth) [reply src/lib/structures/fpa_rounding.ml]: I put a note and I will do it in another PR, maybe after the release.

**Comment by Halbaroth** [comment]
No regression with the following combinations:
- frontend `legacy` on `ae-format`
- frontend `legacy` on `ae-format-fpa` with `--enable-theories fpa` 
- frontend `dolmen` on `ae-format`
- frontend `dolmen` on `ae-format-fpa` with `--enable-theories fpa` 

So this PR is ready for I hope a last check ;)

**Review by bclement-ocp — APPROVED** [review]
:tada:

**Pull request merged**
